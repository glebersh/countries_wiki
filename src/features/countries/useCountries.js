import { useEffect } from 'react'

import { useDispatch, useSelector } from 'react-redux';

import { selectControls } from '../controls/controlsSlice';
import {
  selectCountriesInfo,
  selectVisibleCountries,
  loadCountries
} from './countriesSlice';

export const useCountries = () => {

  const { search, region } = useSelector(selectControls);
  const countries = useSelector(state => selectVisibleCountries(state,
    { search, region }));
  const dispatch = useDispatch();

  const { status, error, numberOfCountries } = useSelector(selectCountriesInfo);

  useEffect(() => {
    if (!numberOfCountries) {
      dispatch(loadCountries());
    }
  }, [numberOfCountries, dispatch]);

  return [countries, { status, error, numberOfCountries }];
};

