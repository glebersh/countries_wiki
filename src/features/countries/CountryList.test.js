import { render, screen, waitFor } from '@testing-library/react';
import { toBeInTheDocument, toHaveBeenCalledWith } from '@testing-library/jest-dom';
import * as countriesHook from './useCountries';
import CountriesList from './CountryList';


const countries = [{
  name: 'Bhutan',
  capital: 'Thimphu',
  region: 'Asia',
  population: 771612,
  flags: {
    svg: 'https://flagcdn.com/bt.svg',
    png: 'https://flagcdn.com/w320/bt.png'
  },
  independent: false
}, {
  name: 'Bolivia (Plurinational State of)',
  capital: 'Sucre',
  region: 'Americas',
  population: 11673029,
  flags: {
    svg: 'https://flagcdn.com/bo.svg',
    png: 'https://flagcdn.com/w320/bo.png'
  },
  independent: false
}]


jest.mock('./useCountries');

const mockedUsedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUsedNavigate,
}));

describe('List component', () => {
  it('List renders if loadingStatus = \'received\'', () => {
    jest.spyOn(countriesHook, 'useCountries').mockImplementation(() => [countries, { status: 'received', error: null, numberOfCountries: 2 }]);
    const list = render(<CountriesList />);
    expect(list.getByTestId('countries-list-container')).not.toBeEmptyDOMElement();
  });

  it('Info snapshot', () => {
    jest.spyOn(countriesHook, 'useCountries').mockImplementation(() => [countries, { status: 'received', error: null, numberOfCountries: 2 }]);
    const list = render(<CountriesList />);
    expect(list).toMatchSnapshot();
  });

  it('List not renders if loadingStatus = \'loading\'', () => {
    jest.spyOn(countriesHook, 'useCountries').mockImplementation(() => [countries, { status: 'loading', error: null, numberOfCountries: 2 }]);
    const list = render(<CountriesList />);
    expect(list.getByTestId('loading-header')).toHaveTextContent('Loading...');
  });


  it('List not renders if error', () => {
    jest.spyOn(countriesHook, 'useCountries').mockImplementation(() => [countries, { status: '', error: true, numberOfCountries: 2 }]);
    const list = render(<CountriesList />);
    expect(list.getByTestId('error-header')).toHaveTextContent('Can\'t fetch data');
  });
});
