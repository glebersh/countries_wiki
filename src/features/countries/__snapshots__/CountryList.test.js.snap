// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component Info snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        data-testid="countries-list-container"
      >
        <section
          class="sc-bcXHqe bglYcP"
        >
          <article
            class="sc-gswNZR iquLsn"
          >
            <img
              alt="Bhutan"
              class="sc-dkrFOg eygKVp"
              src="https://flagcdn.com/w320/bt.png"
            />
            <div
              class="sc-hLBbgP lmooVC"
            >
              <h3
                class="sc-eDvSVe dISRiD"
              >
                Bhutan
              </h3>
              <ul
                class="sc-jSUZER hAhOSo"
              >
                <li
                  class="sc-gKPRtg hRDTNX"
                >
                  <b>
                    Population
                    :
                  </b>
                   
                  771 612
                </li>
                <li
                  class="sc-gKPRtg hRDTNX"
                >
                  <b>
                    Region
                    :
                  </b>
                   
                  Asia
                </li>
                <li
                  class="sc-gKPRtg hRDTNX"
                >
                  <b>
                    Capital
                    :
                  </b>
                   
                  Thimphu
                </li>
              </ul>
            </div>
          </article>
          <article
            class="sc-gswNZR iquLsn"
          >
            <img
              alt="Bolivia (Plurinational State of)"
              class="sc-dkrFOg eygKVp"
              src="https://flagcdn.com/w320/bo.png"
            />
            <div
              class="sc-hLBbgP lmooVC"
            >
              <h3
                class="sc-eDvSVe dISRiD"
              >
                Bolivia (Plurinational State of)
              </h3>
              <ul
                class="sc-jSUZER hAhOSo"
              >
                <li
                  class="sc-gKPRtg hRDTNX"
                >
                  <b>
                    Population
                    :
                  </b>
                   
                  11 673 029
                </li>
                <li
                  class="sc-gKPRtg hRDTNX"
                >
                  <b>
                    Region
                    :
                  </b>
                   
                  Americas
                </li>
                <li
                  class="sc-gKPRtg hRDTNX"
                >
                  <b>
                    Capital
                    :
                  </b>
                   
                  Sucre
                </li>
              </ul>
            </div>
          </article>
        </section>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      data-testid="countries-list-container"
    >
      <section
        class="sc-bcXHqe bglYcP"
      >
        <article
          class="sc-gswNZR iquLsn"
        >
          <img
            alt="Bhutan"
            class="sc-dkrFOg eygKVp"
            src="https://flagcdn.com/w320/bt.png"
          />
          <div
            class="sc-hLBbgP lmooVC"
          >
            <h3
              class="sc-eDvSVe dISRiD"
            >
              Bhutan
            </h3>
            <ul
              class="sc-jSUZER hAhOSo"
            >
              <li
                class="sc-gKPRtg hRDTNX"
              >
                <b>
                  Population
                  :
                </b>
                 
                771 612
              </li>
              <li
                class="sc-gKPRtg hRDTNX"
              >
                <b>
                  Region
                  :
                </b>
                 
                Asia
              </li>
              <li
                class="sc-gKPRtg hRDTNX"
              >
                <b>
                  Capital
                  :
                </b>
                 
                Thimphu
              </li>
            </ul>
          </div>
        </article>
        <article
          class="sc-gswNZR iquLsn"
        >
          <img
            alt="Bolivia (Plurinational State of)"
            class="sc-dkrFOg eygKVp"
            src="https://flagcdn.com/w320/bo.png"
          />
          <div
            class="sc-hLBbgP lmooVC"
          >
            <h3
              class="sc-eDvSVe dISRiD"
            >
              Bolivia (Plurinational State of)
            </h3>
            <ul
              class="sc-jSUZER hAhOSo"
            >
              <li
                class="sc-gKPRtg hRDTNX"
              >
                <b>
                  Population
                  :
                </b>
                 
                11 673 029
              </li>
              <li
                class="sc-gKPRtg hRDTNX"
              >
                <b>
                  Region
                  :
                </b>
                 
                Americas
              </li>
              <li
                class="sc-gKPRtg hRDTNX"
              >
                <b>
                  Capital
                  :
                </b>
                 
                Sucre
              </li>
            </ul>
          </div>
        </article>
      </section>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
